torch/csrc/autograd/python_variable.cpp

```diff
@@ -119,22 +119,46 @@ PyObject * THPVariable_Wrap(Variable var)
 static int THPVariable_traverse(THPVariable *self, visitproc visit, void *arg)
 {
   Py_VISIT(self->backward_hooks);
-  // We don't want to traverse the grad_fn, even if the Variable owns it and the
-  // shared pointer's use count is 1. This is because we would need to treat
-  // the grad_fn as part of the Python state and hold the GIL sometimes when
-  // grad_fn's shared_ptr is copied, otherwise a race condition with the Python
-  // GC could occur. Holding the GIL when the shared_ptr is copied adds
-  // undesirable complexity/overhead.
-  //
-  // When hooks, a Variable, and its grad_fn are involved in a Python reference
-  // cycle, because we're not traversing the grad_fn, the reference cycle will
-  // in fact leak.
-  //
-  // See https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
-  // for more details about the race condition involving traversing the grad_fn
-  // and the python GC.
   const auto& tensor = THPVariable_Unpack(self);
   if (tensor.defined()) {
+    // We don't want to traverse the grad_fn, even if the Variable owns it and the
+    // shared pointer's use count is 1. This is because we would need to treat
+    // the grad_fn as part of the Python state and hold the GIL sometimes when
+    // grad_fn's shared_ptr is copied, otherwise a race condition with the Python
+    // GC could occur. Holding the GIL when the shared_ptr is copied adds
+    // undesirable complexity/overhead.
+    //
+    // When hooks, a Variable, and its grad_fn are involved in a Python reference
+    // cycle, because we're not traversing the grad_fn, the reference cycle will
+    // in fact leak.
+    //
+    // See https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
+    // for more details about the race condition involving traversing the grad_fn
+    // and the python GC.
+
+    // We want a special case though when this THPVariable has the only owning
+    // reference to the Tensor and the Tensor has the only owning reference to
+    // the grad_fn. In such cases, we know that the above case cannot happen
+    // as the only way to access the grad_fn is via this THPVariable which is a
+    // python object and we hold the GIL right now.
+    // In such cases, we can traverse the grad_fn.
+    if (tensor.use_count() == 1) {
+      auto autograd_meta = torch::autograd::impl::get_autograd_meta(tensor);
+      if (autograd_meta) {
+        // Do NOT call grad_fn() here as that might trigger a recompute
+        const auto& grad_fn = autograd_meta->grad_fn_;
+        if (grad_fn && grad_fn.use_count() == 1) {
+          // All Node can have a pyobj
+          Py_VISIT(grad_fn->pyobj());
```

Comment:
How come the visit on `grad_fn->pyobj()` doesn't induce a visit on `py_node_fn->obj`?

Reply:
Because these are two independent fields.
And py_node_fn has both a `pyobj_` field and an `obj` field.
As far as I know, the `pyobj_` field is never used for the PyNodes though as there is special handling in the `pyobj()` function to read the `obj` field for them instead of `pyobj_`.